packages/excalidraw/package.json

```diff
@@ -0,0 +1,27 @@
+{
+  "name": "@prairielearn/excalidraw",
+  "version": "0.0.1",
+  "type": "module",
+  "main": "dist/index.js",
+  "sideEffects": false,
+  "files": [
+    "dist"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/PrairieLearn/PrairieLearn.git",
+    "directory": "packages/excalidraw"
+  },
+  "scripts": {
+    "build": "node esbuild.config.js"
+  },
+  "//": "Note: We need https://github.com/excalidraw/excalidraw/pull/8030, which is not yet released. Undo the pinning when a new excalidraw release is made",
```

Comment:
AFAICT, the prerelease builds include a ton of dependencies in `package.json`, whereas the "real" releases don't. Installing `@excalidraw/excalidraw@latest` on `master` resulted in a much, much smaller diff to `yarn.lock`. Ultimately, we want that, as we _don't_ want to install a bunch of unnecessary dependencies that would bloat `node_modules`. Given that, I think it'll be a hard requirement to see a new release of Excalidraw before we can merge this PR.

Reply:
The latest excalidraw release 0.17.x is fundamentally different in how it's built compared to the unreleased v0.18.x (which oddly is called 0.17.1-commit). 0.17.x recommends using the prebuilt UMD dist and requires statically serving assets (fonts, i18n) but 0.18.x switches to ESM and we control how it's packaged with esbuild. I don't think it would make a difference if we wait for 0.18.x because we are no longer using the prebuilt stuff.

Do you have any specific instances of unnecessary dependencies you see on the diff? Most packages (not all, but most) look like they are necessary at runtime.

I would like to understand the problem here - are we worried about disk space or incremental build times, or that the yarn.lock diff is too large?

Reply:
My concern is not necessarily that the `yarn.lock` diff is so large; that's just a symptom of the fact that excalidraw transitively pulls in a metric ton of stuff. I opened https://github.com/PrairieLearn/PrairieLearn/pull/9928 so that our Docker image size analysis could run (it doesn't run on PRs from forks). This PR results in a 2% increase in our compressed image size, or 23MB. Uncompressed, there's now an addition _104MB_ of stuff installed:

```
$ docker inspect -f "{{ .Size }}" prairielearn/prairielearn:latest
3499438952
$ docker inspect -f "{{ .Size }}" prairielearn/prairielearn:1fb0e8f4879f5a4f1e8f2542f45c993553153ce9
3603780841
```

This doesn't even include `react` or `react-dom` yet.

I think we avoided much of this cost increase when excalidraw was publishing a pre-built UMD bundle. Do you know if they have any plans of publishing a separate package that's bundled the same way as the old one?

Reply:
Just again raising [js-draw](https://www.npmjs.com/package/js-draw) as an alternative. Its dependencies seem to be way simpler. As a heuristic I did a `yarn add js-draw` locally and saw an increase of ~8-9MB (uncompressed) in my local dir. It also generates an SVG as output and is able to open it as input, so might also have the advantage to make it easier to swap engines in the future.

Reply:
Unfortunately this late in development I can't justify paying for more hours to switch the backend again and starting over. 

Is this truly a deal breaker?

Reply:
I thought we wanted to move away from using UMD - In an OH, you suggested moving to our own bundling setup so that we avoid registering a global `window.React` to the version of React I use with Excalidraw. Along a similar line of thought, I figured we wanted to not register a `window.ExcalidrawLib` and make it an ESM.

> Do you know if they have any plans of publishing a separate package that's bundled the same way as the old one?

They still do UMD but switching to UMD implies we no longer control the entire build process and serve EXCALIDRAW_STATIC_ASSETS manually. Switching to UMD _and_ depending on their static assets directory to be served is IMO objectively worse compared to a bundle size increase.

> I think we avoided much of this cost increase when excalidraw was publishing a pre-built UMD bundle.

I don't think I quite understand the issue so perhaps let me rephrase my question - are we worried that (A) installing excalidraw increases the size of `node_modules` or that (B) the static assets served at runtime increase in size?

(A) It doesn't matter if we use UMD or ESM, the node_modules folder _will_ go up by 50MB because excalidraw distributes bundles for dev/prod x umd/esm.

(B) UMD `@excalidraw/excalidraw/dist/browser/prod` is 6.3MB and our ESM bundle is 6.5MB. Once again, UMD or ESM does not matter. I haven't looked into tree shaking opportunities but we could cut down the bundle size if we decide to drop all languages except en-US. Moreover our ESM is chunked so we aren't loading 6MB worth of JS in one go. I think the default setup takes around 1MB the last time I checked.

We could request excalidraw devs to split the dist bundle into distinct packages like you suggest. Do you consider this a blocker?

I have two alternative proposals -

- fork excalidraw and reduce dist/ size (we are clear we don't want to maintain forks, so this is a no)
- I do some hackery to invoke npm using esbuild and maintain a separate lockfile for prairielearn/excalidraw. Pros: 6MB bump in node_modules size; Cons: Poor maintainability

---

js-draw is worse feature-wise compared to excalidraw (based on like 5 minutes of trying it out)

- flowcharting is worse because arrows don't track objects
- the default layout takes a bunch of more clicks to change tools / colors (we could customize this however)
- no way of changing properties of a shape that's already drawn
- the copy-pasting is awkward and there is no grouping or layering

excalidraw also provides a component library (currently unused) which could be used in the future - if I was doing an org-chem course and I needed benzene, I'd just pick the shape from a chem library.

---

Aside: I think node_modules generally could use some tree shaking for prod containers because we know which node_modules we want served at runtime, statically. Happy to discuss IRL if you're interested.

Reply:
> I thought we wanted to move away from using UMD - In an OH, you suggested moving to our own bundling setup so that we avoid registering a global window.React to the version of React I use with Excalidraw.

Yes, but this is orthogonal to how Excalidraw itself bundles their own code. We could just as easily consume their UMD code into our own bundle.

> I don't think I quite understand the issue so perhaps let me rephrase my question - are we worried that (A) installing excalidraw increases the size of node_modules or that (B) the static assets served at runtime increase in size?

(A) is the primary concern. (B) _would_ be a concern if it happened, but I don't think that's likely to be the case; there's no direct correlation between size of `node_modules` and size of bundled assets and I trust that our bundling setup will perform similarly to theirs.

> (A) It doesn't matter if we use UMD or ESM, the node_modules folder will go up by 50MB because excalidraw distributes bundles for dev/prod x umd/esm.

This is a good point, I admittedly didn't check the current `@latest` version but indeed, the package itself includes a boatload of built files (only ~5MB of which we actually care about):

<img width="783" alt="Screenshot 2024-05-31 at 15 55 11" src="https://github.com/PrairieLearn/PrairieLearn/assets/1476544/375e8a11-c55e-446a-b9dd-4918ab081efe">

So even if they weren't changing how they publish their packages, we'd still be dragging in quite a lot of files that we don't use.

Is any of this a blocker? Not necessarily, but I've fought tooth and nail to avoid dramatic increasing the size of our images and even to shrink them, so a (relatively) large increase like this gives me pause.

Another possible option is that we could do all of this bundling in a separate repo (`PrairieLearn/excalidraw-builds`?). While we typically like keeping as much as possible in one repo to keep management simpler, this might be a case where the tradeoff is worth it. This separate repo could declare `@excalidraw/excalidraw` in `devDependencies`, and it would publish with _just_ the built files we need. This keeps their dependency bloat far away from PrairieLearn itself.

I don't know if I can make the call on the right way forward on my own. Let me run this all by some of the other core devs.

Thank you for your patience. We're doing something pretty novel here and working out the best way to do things as we go!

Reply:
After discussion with some of the other core devs, we've decided that building + publishing a new package from a separate repository will be the best course of action. Given how infrequently Excalidraw seems to change, this shouldn't pose a large maintenance burden going forwards.

It'll be on me to set up this repository and get auto-publishing working. I'll create the repository now so that you can start lifting over the code from this PR. Once we have something that's publishable, I'll work on getting the publishing tooling set up.

Reply:
I've created the repository here: https://github.com/PrairieLearn/excalidraw-builds. Let me know if you'd like me to take on some additional repository prep like getting Yarn and such set up. If you have questions about this approach, please ask!

Reply:
I've asked @firasm about the (additional) allocation of hours. I'll post an update once we finalize the details.

Reply:
Sounds good. We can take this over the finish line for you if the additional hours will be an issue.

Reply:
@nwalters512 Thanks for making the new repo. Publishing a new package there should only result in minor changes to this PR - removing the esbuild code and renaming `@prairielearn/excalidraw` to the npm-published version. Does it make sense that we continue reviewing the remaining changes in this PR? Knowing the scope of the requested changes (besides the new excalidraw package) would help me plan my time. If you could start a thread, one each for the remaining changes you'd like addressed in this PR, that'd be great.

Reply:
Can do! I should be able to get to that either tomorrow or Monday. I don't anticipate needing a lot of changes.

Reply:
@nishanthkarthik had the great idea to [ask](https://github.com/excalidraw/excalidraw/issues/8104) in the Excalidraw repo about this

Their response: https://github.com/excalidraw/excalidraw/issues/8104#issuecomment-2153801914 👀

Reply:
@nwalters512 I've created https://github.com/PrairieLearn/excalidraw-builds/pull/1. Are you thinking of setting up auto-publish through github actions?

Reply:
Auto-publishing was implemented in https://github.com/PrairieLearn/excalidraw-builds/pull/2; see the readme for details.

Reply:
![image](https://github.com/PrairieLearn/PrairieLearn/assets/7759435/efcc2afb-0f15-4c29-b950-33870409e11f)


Nice work!

I'll fix this PR tomorrow (or by Friday)

Reply:
fixed, resolving